### DISCUSSION TITLE
Welcome to Obsidian-RunJS Discussions!

### DISCUSSION THREAD

**Discussion by eoureo** [eoureo/obsidian-runjs#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Share codes for RunJS.
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by gapmiss** [comment]
Greetings~

Thank you for your contribution to Obsidian development. This plugin scratches a few itches for myself.

I have a couple of use-cases and would appreciate some examples of whether or not RunJS can accomplish them.

How can one import/require from outside the local file system?

e.g.:

```js
const { default: {mermaidAPI} } = await import('https://cdn.jsdelivr.net/npm/mermaid@10/dist/mermaid.esm.min.mjs')

mermaidAPI.initialize({ startOnLoad: false });

// ------

const ctx = contentEl.createEl("canvas", null);

await import("https://cdnjs.cloudflare.com/ajax/libs/Chart.js/4.2.1/chart.umd.js");
```


Secondly, how could one `render` output to the note that contains the `RunJS` code?

e.g.

```js
let el = contentEl.createEl("div", null);
const graphDefinition = `
timeline
    title History of Social Media Platform
    2002 : LinkedIn
    2004 : Facebook
         : Google
    2005 : Youtube
    2006 : Twitter
`

const {svg} = await mermaidAPI.render("graphDiv", graphDefinition);
el.innerHTML = svg;
```

The above example are derived for the Dataview plugin and can be found here:

- [Use ChartJS with DataView - Share & showcase - Obsidian Forum](https://forum.obsidian.md/t/use-chartjs-with-dataview/58752)
- [Use Updated Version of Mermaid JS with DataViewJS - Share & showcase - Obsidian Forum](https://forum.obsidian.md/t/use-updated-version-of-mermaid-js-with-dataviewjs/58748)

Thank you

**Reply by eoureo** [reply]
Hi!
Thank you for your nice comment.
As a non-native English speaker, I kindly request you to consider my limited English proficiency while reading my text.

RunJS is actually a plugin I personally created to make handling Obsidian and note files easier using JavaScript. That being said, it still has some room for improvement. I'm sharing the source code with a bit of doubt, wondering if it's helpful to others and if it stands out from other plugins.

Now, let me respond to your comment.

* Import
Since RunJS is still in the early stages of development, I haven't added the feature to import remote external modules just yet. I've put that on hold to evaluate whether it's something I can include later. However, after taking a look at your feedback, making some tweaks to the source code, and running tests, I've confirmed that importing is indeed possible.
I'll need to check the constraints imposed by the Obsidian Plugin (regarding security issues).
I don't think it should be a major problem. If that's the case, I'll go ahead and add the external import function.

* Render
At the moment, RunJS doesn't have its own rendering function. Your comment has sparked some great ideas for me.

Once again, thank you so much for your comment.
^____^

**Comment by nqm0711** [comment]
Hi, thank you for your plugin that it can extend the flexibility of Obsidian so much, I'm very appreciate that.

Could you please add some extra description about event handler such as which one should be selected in which case, I'm not a developer and just a beginner user of Obsidian and I actually don't know which will be considered as active-leaf-change and which will be editor change.

I just use chatgpt to wrote a script based on my description to create button and change the display on click. I used RunJS to run that script and it worked perfectly when I run it manually.

But now I want it to run on startup (I'm already added it to the "auto start" setting), run everytime I change between preview and reading mode, and run everytime I modify the note content ( or if possible run everytime I add/modify some elements with a specified class) but I don't know which one to select yet. Please kindly provide some description for these options.

Many thanks
